pkg/util/informermanager/federatedinformermanager.go

```diff
@@ -0,0 +1,328 @@
+package informermanager
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"fmt"
+	"sync"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	fedcorev1a1 "github.com/kubewharf/kubeadmiral/pkg/apis/core/v1alpha1"
+	fedcorev1a1informers "github.com/kubewharf/kubeadmiral/pkg/client/informers/externalversions/core/v1alpha1"
+	fedcorev1a1listers "github.com/kubewharf/kubeadmiral/pkg/client/listers/core/v1alpha1"
+	"github.com/kubewharf/kubeadmiral/pkg/controllers/util"
+)
+
+type federatedInformerManager struct {
+	lock sync.RWMutex
+
+	started bool
+
+	clientGetter    ClusterClientGetter
+	ftcInformer     fedcorev1a1informers.FederatedTypeConfigInformer
+	clusterInformer fedcorev1a1informers.FederatedClusterInformer
+
+	eventHandlerGenerators []*EventHandlerGenerator
+	clusterEventHandler    []*ClusterEventHandler
+
+	clients                     map[string]dynamic.Interface
+	connectionMap               map[string][]byte
+	informerManagers            map[string]InformerManager
+	informerManagersCancelFuncs map[string]context.CancelFunc
+
+	queue  workqueue.Interface
+	logger klog.Logger
+}
+
+func NewFederatedInformerManager(
+	clientGetter ClusterClientGetter,
+	ftcInformer fedcorev1a1informers.FederatedTypeConfigInformer,
+	clusterInformer fedcorev1a1informers.FederatedClusterInformer,
+) FederatedInformerManager {
+	manager := &federatedInformerManager{
+		lock:                        sync.RWMutex{},
+		started:                     false,
+		clientGetter:                clientGetter,
+		ftcInformer:                 ftcInformer,
+		clusterInformer:             clusterInformer,
+		eventHandlerGenerators:      []*EventHandlerGenerator{},
+		clusterEventHandler:         []*ClusterEventHandler{},
+		clients:                     map[string]dynamic.Interface{},
+		connectionMap:               map[string][]byte{},
+		informerManagers:            map[string]InformerManager{},
+		informerManagersCancelFuncs: map[string]context.CancelFunc{},
```

Comment:
Would it be better if we propagate the context passed into `Run` all the way and use it to start the `InformerManager`s? We can do the same for the stop channels in `InformerManager`.

Reply:
That's a good idea

Reply:
If we do so, we don't need to store the `CancelFunc`s in the informer manager.

Reply:
We might still have to, refer to the conversation below about removing `CancelFuncs`